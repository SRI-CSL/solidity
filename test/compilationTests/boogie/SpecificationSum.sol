pragma solidity ^0.4.23;

/**
 * @notice invariant __verifier_sum(user_balances) == total_supply
 */
contract SomeContract {
    int total_supply;
    mapping(address=>int) user_balances;

    function deposit(int amount) public {
        total_supply += amount;
        user_balances[msg.sender] += amount;
    }

    function withdraw() public {
        total_supply -= user_balances[msg.sender];
        user_balances[msg.sender] = 0;
    }

    function withdraw_incorrect() public {
        total_supply -= user_balances[msg.sender];
        user_balances[msg.sender] = 12345;
    }
}
