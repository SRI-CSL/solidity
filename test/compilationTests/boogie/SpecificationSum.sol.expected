Verifying SomeContract::deposit: OK
Verifying SomeContract::withdraw: OK
Verifying SomeContract::withdraw_incorrect: ERROR
 - Source SpecificationSum.sol, line 20, col 5: Invariant '__verifier_sum_uint8(user_balances) == total_supply' might not hold at end of function.
